app/channel.go

```diff
@@ -322,69 +322,70 @@ func (a *App) CreateChannel(channel *model.Channel, addMember bool) (*model.Chan
 
 func (a *App) GetOrCreateDirectChannel(userId, otherUserId string) (*model.Channel, *model.AppError) {
 	channel, nErr := a.Srv().Store.Channel().GetByName("", model.GetDMNameFromIds(userId, otherUserId), true)
-	if nErr != nil {
-		var nfErr *store.ErrNotFound
-		if errors.As(nErr, &nfErr) {
-			var err *model.AppError
-			channel, err = a.createDirectChannel(userId, otherUserId)
-			if err != nil {
-				if err.Id == store.ChannelExistsError {
-					return channel, nil
-				}
-				return nil, err
-			}
-
-			a.WaitForChannelMembership(channel.Id, userId)
+	if nErr == nil {
+		return channel, nil
+	}
+	var nfErr *store.ErrNotFound
+	if !errors.As(nErr, &nfErr) {
+		return nil, model.NewAppError("GetOrCreateDirectChannel", "web.incoming_webhook.channel.app_error", nil, nErr.Error(), http.StatusInternalServerError)
+	}
 
-			a.InvalidateCacheForUser(userId)
-			a.InvalidateCacheForUser(otherUserId)
+	channel, err := a.createDirectChannel(userId, otherUserId)
+	if err != nil {
+		if err.Id == store.ChannelExistsError {
+			return channel, nil
+		}
+		return nil, err
+	}
 
-			if pluginsEnvironment := a.GetPluginsEnvironment(); pluginsEnvironment != nil {
-				a.Srv().Go(func() {
-					pluginContext := a.PluginContext()
-					pluginsEnvironment.RunMultiPluginHook(func(hooks plugin.Hooks) bool {
-						hooks.ChannelHasBeenCreated(pluginContext, channel)
-						return true
-					}, plugin.ChannelHasBeenCreatedId)
-				})
-			}
+	a.WaitForChannelMembership(channel.Id, userId)
 
-			message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_DIRECT_ADDED, "", channel.Id, "", nil)
-			message.Add("teammate_id", otherUserId)
-			a.Publish(message)
+	a.InvalidateCacheForUser(userId)
+	a.InvalidateCacheForUser(otherUserId)
 
-			return channel, nil
-		}
-		return nil, model.NewAppError("GetOrCreateDirectChannel", "web.incoming_webhook.channel.app_error", nil, nErr.Error(), http.StatusInternalServerError)
+	if pluginsEnvironment := a.GetPluginsEnvironment(); pluginsEnvironment != nil {
+		a.Srv().Go(func() {
+			pluginContext := a.PluginContext()
+			pluginsEnvironment.RunMultiPluginHook(func(hooks plugin.Hooks) bool {
+				hooks.ChannelHasBeenCreated(pluginContext, channel)
+				return true
+			}, plugin.ChannelHasBeenCreatedId)
+		})
 	}
+
+	message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_DIRECT_ADDED, "", channel.Id, "", nil)
+	message.Add("teammate_id", otherUserId)
+	a.Publish(message)
+
 	return channel, nil
 }
 
 func (a *App) createDirectChannel(userId string, otherUserId string) (*model.Channel, *model.AppError) {
-	uc1 := make(chan store.StoreResult, 1)
-	uc2 := make(chan store.StoreResult, 1)
-	go func() {
-		user, err := a.Srv().Store.User().Get(userId)
-		uc1 <- store.StoreResult{Data: user, NErr: err}
-		close(uc1)
-	}()
-	go func() {
-		user, err := a.Srv().Store.User().Get(otherUserId)
-		uc2 <- store.StoreResult{Data: user, NErr: err}
-		close(uc2)
-	}()
+	users, err := a.Srv().Store.User().GetMany([]string{userId, otherUserId})
```

Comment:
This looks like a great idea, but it is not using the users cache and the get method does it, probably this is less performant in the real environment. We can load test this to verify it. Anyway I would suggest measure it, or add the cache usage to the new GetMany method.

Reply:
+1 to this. Let's add the `GetMany` method to the cache layer which gets the values from multiple ids.

Reply:
oh yeah, completely forgot about the cache. Will added to the cache.

Reply:
amended in https://github.com/mattermost/mattermost-server/pull/16819/commits/3ba1726037e26841ecf2b9ca8afd5e6d8a289c7a